# Functions:

## Normal Functions:

### Write a Python function called "sum" that takes two numbers as parameters and returns their sum.

In [1]:
def sum_ab(a,b):
    c = a + b
    return c

sum_ab(8,10)

18

### Write a function called "greatest_number" that takes a list of numbers as a parameter and returns the largest number in the list.

In [2]:
def greatest_number(list):
    list.sort(reverse=True)
    number = list[0]
    return number

list1=[1,2,3,4,5,6,7,8,9,10,80,120]

greatest_number(list1)

120

### Write a function called "average" that takes a list of numbers as a parameter and returns the arithmetic mean of the elements in the list.

In [3]:
list1 = [1,2,3,4,5,6,7,8,9]
def avarege(list):
    size_list =len(list)
    sum_list = sum(list)
    num_avarage = sum_list/size_list
    return num_avarage

avarege(list1)

5.0

### Write a function called "even_or_odd" that takes a number as a parameter and returns "even" if the number is even, or "odd" if the number is odd.

In [4]:
def even_or_odd(number):
    if number % 2 == 0:
        return "even"
    else:
        return "odd"

result = even_or_odd(5)
print("It is", result)

It is odd


### Write a function called "factorial" that takes an integer as a parameter and returns the factorial of that number. The factorial of a number is the product of all integers from 1 to the number itself.

In [5]:
def factorial(number):
    if number < 0:
        return "Factorial is not defined for negative numbers."
    elif number == 0 or number == 1:
        return 1
    else:
        result = 1
        for i in range(2, number + 1):
            result *= i
        return result
num = 6
result = factorial(num)
print('The factorial of', num, 'is:', result)

The factorial of 6 is: 720


## Lambda function:

### Lambda function that returns the square of a number:

In [6]:
square_number = (lambda x: x * x)(5)
print(square_number)

25


### Lambda function that checks if a number is even:

In [7]:
check_even = (lambda x : 'even number' if x % 2 == 0 else 'odd number')(10)
print(check_even)

even number


### Lambda function that returns the sum of two numbers:

In [8]:
sum_tow_numbers = (lambda a, b: a+b)(5,10)
print(sum_tow_numbers)

15


### Lambda function that returns the absolute value of a number:

In [9]:
absolute_value = (lambda x: abs(x))(-5)
print(absolute_value)

5


### Lambda function that checks if a word is a palindrome (reads the same backwards):

In [10]:
is_palindrome = (lambda x: x==x[::-1])
print(is_palindrome('ana'))

True


### Lambda function that returns the double of each element in a list:

In [11]:
list_x = [1, 2, 3, 4, 5]

lambda_double = lambda list: [element * 2 for element in list]

print(lambda_double(list_x))

[2, 4, 6, 8, 10]


## MAP function:

### Exercise 1: Convert Temperatures. In this exercise, we are going to convert a list of temperatures in degrees Celsius to Fahrenheit using the map() function. List of temperatures in Celsius: temperatures_celsius = [25, 30, 15, 20, 10]

In [12]:
list_temp = [20,10,-5,30,0]

convert_celsius_to_fahrenheit = list(map(lambda x: (x * 9/5) + 32, list_temp))

print(convert_celsius_to_fahrenheit)

[68.0, 50.0, 23.0, 86.0, 32.0]


### Exercise 2: Calculating Word Length In this exercise, we will calculate the length of each word in a list of words using the map() function. Wordlist, words = ["Python", "is", "a", "language", "of", "programming"]

In [13]:
list_words = ["Python", "is", "a", "language", "of", "programming"]

calculing_word_lengh = list(map(lambda word: len(word), list_words))

print(calculing_word_lengh)

[6, 2, 1, 8, 2, 11]


### Exercise 3: Converting Numbers to Strings In this exercise, we are going to convert a list of integers into a list of strings using the map() function.List of integers numbers = [10, 20, 30, 40, 50]

In [14]:
integers_numbers = [10, 20, 30, 40, 50]

converting_numbers_to_strings = list(map(lambda inter: str(inter), integers_numbers))

print(converting_numbers_to_strings)

['10', '20', '30', '40', '50']


### Exercise 4: Raising Numbers to Powers In this exercise, we will raise each number in a list to a power using the map() function.
### List of numbers numbers = [2, 3, 4, 5], Desired power power = 3

In [15]:
list_numbers = [2,3,4,5]

numbers_power_3 = list(map(lambda numb: numb**3, list_numbers))

print(numbers_power_3)

[8, 27, 64, 125]


## Map function with Lambda:

### Exercise 1: Folding the Elements of a List
### Description: Given a list of numbers, create a new list containing each doubled number.
### List of numbers: [1, 2, 3, 4, 5]
### Answer: [2, 4, 6, 8, 10]

In [16]:
list_numbers = [1,2,3,4,5]

double_numbers = list(map(lambda numb: numb*2,list_numbers))

print(double_numbers)

[2, 4, 6, 8, 10]


### Exercise 2: Convert each element to a string and add a suffix
### Description: Given a list of numbers, convert each element to a string and add the suffix "kg".
### List of numbers: [10, 20, 30, 40, 50]
### Answer: ['10kg', '20kg', '30kg', '40kg', '50kg']

In [17]:
list_of_numbers = [10,20,30,40,50]

suffix_kg = list(map(lambda inter: str(inter)+'kg',list_of_numbers))

print(suffix_kg)

['10kg', '20kg', '30kg', '40kg', '50kg']


### Exercise 3: Filter only even numbers
### Description: Given a list of numbers, filter only even numbers.
### List of numbers: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
### Answer: [2, 4, 6, 8, 10]

In [18]:
list_of_numbers = [1,2,3,4,5,6,7,8,9,10]

numeros_pares = list(filter(None, map(lambda x: x if x % 2 == 0 else None, list_of_numbers)))

print(numeros_pares)

[2, 4, 6, 8, 10]


In [19]:
list_of_numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
numeros_pares = list(map(lambda x: x if x % 2 == 0 else None, list_of_numbers))
numeros_pares = list(filter(lambda x: x is not None, numeros_pares))
print(numeros_pares)

[2, 4, 6, 8, 10]


### Exercise 4: Capitalize each word
### Description: Given a list of words, make each word uppercase.
### Word list: ["python", "is", "fun"]
### Answer: ["PYTHON", "IS", "FUN"]

In [20]:
word_list = ["python", "is", "fun"]

right_word_list = list(map(lambda word: word.capitalize(),word_list))

print(right_word_list)

['Python', 'Is', 'Fun']


### Exercise 5: Calculating the product of two numbers in two different lists
### Description: Given two lists of numbers, calculate the product between the corresponding elements of the two lists.
### List of numbers 1: [1, 2, 3, 4, 5]
### List of numbers 2: [10, 20, 30, 40, 50]
### Answer: [10, 40, 90, 160, 250]

In [21]:
list_of_numbers_1 = [1, 2, 3, 4, 5]
list_of_numbers_2 = [10, 20, 30, 40, 50]

answer = list(map(lambda x: x[0] + x[1], zip(list_of_numbers_1, list_of_numbers_2)))

print(answer)

[11, 22, 33, 44, 55]


# Data structure:

## Lists of lists

### Exercise 1: Sum elements of inner lists
### Description: Given a list of lists, sum the corresponding elements of each inner list and return a new list with the sums.
### List of lists: [[1, 2, 3], [4, 5, 6], [7, 8, 9]]
### Answer: [6, 15, 24]

In [22]:
def sum_lists(list):
    """
    sums the elements of inner lists.

    Args:
        list_of_list: A list of lists

    Returns:
        A list with them sums of the inner lists.
    """

    sums = []
    for inner_list in list_of_lists:
        sum = 0
        for element in inner_list:
            sum += element
        sums.append(sum)
    return sums

list_of_lists = [[1,2,3],[4,5,6],[7,8,9]]
sums = sum_lists(list_of_lists)
print(sums)

[6, 15, 24]


### Exercise 2: Compute the product of elements of inner lists
### Description: Given a list of lists, calculate the product between the corresponding elements of each internal list and return a new list with the products.
### List of lists: [[2, 3, 4], [1, 2, 3], [4, 5, 6]]
### Answer: [24, 6, 120]

In [23]:
def product_list(list_of_lists):
  products = []
  for inter_list in list_of_lists:
    product = 1
    for element in inter_list:
      product *= element
    products.append(product)
  return products
list_of_lists = [[2, 3, 4], [1, 2, 3], [4, 5, 6]]
products = product_list(list_of_lists)
print(products)

[24, 6, 120]


### Exercise 3: Concatenating Strings from Internal Lists
### Description: Given a list of lists, concatenate the strings of each internal list and return a new list with the concatenated strings.
### List of lists: [["Hello", "World"], ["Python", "is"], ["Cool", "!"]]
### Response: ['HelloWorld', 'Pythois', 'Nice!']

In [24]:
concatenated_strings = []
def concatenate_strings(list):
    for inter_list in list:
        concatenated_string = ""
        for element in inter_list:
            concatenated_string += element
        concatenated_strings.append(concatenated_string)
    return concatenated_strings

list_of_lists = [['hello','world'],['python','is'],['cool','!']]
concatenated_strings = concatenate_strings(list_of_lists)
print(concatenated_strings)

['helloworld', 'pythonis', 'cool!']


### Exercise 4: Average values in inner lists
### Description: Given a list of lists, average the values of each inner list and return a new list with the corresponding means.
### List of lists: [[10, 20, 30], [5, 15, 25], [8, 12, 16]]
### Answer: [20.0, 15.0, 12.0]

In [25]:
avg_list = []
def avg_lists(list):
    for inter_list in list:
        sum = 0
        for element in inter_list:
            sum += element
        avg = sum/len(inter_list)    
        avg_list.append(avg)    
    return avg_list

list_of_lists = [[10, 20, 30], [5, 15, 25], [8, 12, 16]]
avg_list = avg_lists(list_of_lists)
print(avg_list)

[20.0, 15.0, 12.0]


## Tuple Lists

### Exercise 1: Calculating the Area of Rectangles

### Description: Given a list of tuples, where each tuple contains the base and height of a rectangle, calculate the area of each rectangle and return a new list with the corresponding areas.

### List of tuples: [(2, 4), (3, 5), (6, 2)]

### Answer: [8, 15, 12]

In [26]:
list_of_tuples = [(2,4),(3,5),(6,2)]
def area_rectangle(list):
  # Create a list to store the areas of the rectangles.
  list_area = []
  # Iterate through the list of tuples.
  for index, inner_tuple in enumerate(list):
    # Create a new list to store the base and height of the rectangle.
    area = []
    # Iterate through the tuple and add the base and height to the `area` list.
    first_element = inner_tuple[0]
    second_element = inner_tuple[1]
    area.append(first_element)
    area.append(second_element)
    # Calculate the area of the rectangle and append it to the `list_area` list.
    area_of_rectangle = first_element * second_element
    list_area.append(area_of_rectangle)
  # Return the `list_area` list.
  return list_area
answer = area_rectangle(list_of_tuples)
print(answer)

[8, 15, 12]


### Exercise 2: Calculating the Product of Elements of Tuples

### Description: Given a list of tuples, where each tuple contains three numbers, calculate the product between the elements of each tuple and return a new list with the products.

### List of tuples: [(2, 3, 4), (1, 2, 3), (4, 5, 6)]

### Answer: [24, 6, 120]

In [30]:
def product_of_elements(tuples):
  """
  Calculates the product of elements of tuples.

  Args:
    tuples: A list of tuples, where each tuple contains three numbers.

  Returns:
    A new list with the products of the elements of each tuple.
  """

  products = []
  for tuple in tuples:
    product = 1
    for number in tuple:
      product *= number
    products.append(product)
  return products


if __name__ == "__main__":
  tuples = [(2, 3, 4), (1, 2, 3), (4, 5, 6)]
  products = product_of_elements(tuples)
  print(products)

[24, 6, 120]


### Exercise 3: Concatenate Strings of Tuples

### Description: Given a list of tuples, where each tuple contains two strings, concatenate the strings of each tuple and return a new list with the concatenated strings.

### List of tuples: [("Hello", "World"), ("Python", "yeah"), ("Cool", "!")]

### Response: ['HelloWorld', 'Pythois', 'Nice!']

In [28]:
list_of_tuples = [("Hello", "World"), ("Python", "yeah"), ("Cool", "!")]
def string_tuples(list):
    list_concatenated = []
    for index, inner_tuple in enumerate(list):
        concatenated_strings = inner_tuple[0] + inner_tuple[1]
        list_concatenated.append(concatenated_strings)
    return list_concatenated
answer = string_tuples(list_of_tuples)
print(answer)

['HelloWorld', 'Pythonyeah', 'Cool!']


### Exercise 4: Calculate the average of values in tuples

### Description: Given a list of tuples, where each tuple contains four numeric values, average the values of each tuple and return a new list with the corresponding averages.

### List of tuples: [(10, 20, 30, 40), (5, 15, 25, 35), (8, 12, 16, 20)]

### Answer: [25.0, 20.0, 14.0]

In [29]:
def calculate_average_of_tuples(tuples_list):
    averages = []
    for tpl in tuples_list:
        average = sum(tpl) / len(tpl)
        averages.append(average)
    return averages

# Example usage
tuples_list = [(10, 20, 30, 40), (5, 15, 25, 35), (8, 12, 16, 20)]
averages = calculate_average_of_tuples(tuples_list)
print(averages)


[25.0, 20.0, 14.0]


## List comprehension:

### Exercise 1 (Beginner):
### Write a list comprehension that generates a list of the first 10 even numbers.
### Output: [2, 4, 6, 8, 10, 12, 14, 16, 18, 20]

In [4]:
alphabet = [chr(x) for x in range(ord('a'), ord('z')+1)]
alphabet

['a',
 'b',
 'c',
 'd',
 'e',
 'f',
 'g',
 'h',
 'i',
 'j',
 'k',
 'l',
 'm',
 'n',
 'o',
 'p',
 'q',
 'r',
 's',
 't',
 'u',
 'v',
 'w',
 'x',
 'y',
 'z']

### Exercise 2 (Beginner):
### Write a list comprehension that generates a list of the squares of numbers from 1 to 10.

### Input: None
### Output: [1, 4, 9, 16, 25, 36, 49, 64, 81, 100]

In [3]:
list = [x for x in range(1,10)]
list

[1, 2, 3, 4, 5, 6, 7, 8, 9]

### Exercise 3 (Intermediate):
### Write a list comprehension that filters out the vowels from a given word.

### Input: word = "hello"
### Output: ['h', 'l', 'l']

In [9]:
word = 'hello'
vowels = ['a','e','i','o','u']
filters_out_vowels = [ ch for ch in word if ch.lower() not in vowels]
filters_out_vowels

['h', 'l', 'l']

### Exercise 4 (Intermediate):
### Write a list comprehension that combines two lists by taking the element at the same index from each list.

### Input: list1 = [1, 2, 3, 4], list2 = ['a', 'b', 'c', 'd']
### Output: [(1, 'a'), (2, 'b'), (3, 'c'), (4, 'd')]

In [32]:
list1 = [1,2,3,4]
list2 = ['a','b','c','d']
combined_list = [(x,y) for x, y in zip(list1, list2)]
print(combined_list)

[(1, 'a'), (2, 'b'), (3, 'c'), (4, 'd')]


### Exercise 5 (Advanced):
### Write a list comprehension that generates a list of prime numbers between 1 and 100.

### Input: None
### Output: [2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97]

In [34]:
prime_numbers = [x for x in range(2, 100) if all(x % i for i in range(2, x))]
prime_numbers

[2,
 3,
 5,
 7,
 11,
 13,
 17,
 19,
 23,
 29,
 31,
 37,
 41,
 43,
 47,
 53,
 59,
 61,
 67,
 71,
 73,
 79,
 83,
 89,
 97]

## Dict comprehension

### Exercise 1 (Beginner):
### Write a dictionary comprehension that generates a dictionary mapping numbers from 1 to 5 to their squares.

### Input: None
### Output: {1: 1, 2: 4, 3: 9, 4: 16, 5: 25}

In [41]:
dictionary = {x: x ** 2 for x in range(1,5)}
print(dictionary)
list = [x ** 2 for x in range(1,5)]
print(list)

{1: 1, 2: 4, 3: 9, 4: 16}
[1, 4, 9, 16]


### Exercise 2 (Beginner):
### Write a dictionary comprehension that swaps the keys and values of a given dictionary.

### Input: original_dict = {'a': 1, 'b': 2, 'c': 3}
### Output: {1: 'a', 2: 'b', 3: 'c'}

In [43]:
original_dict = {'a': 1, 'b': 2, 'c': 3}
swap_keys = {value: key for key, value in original_dict.items()}
swap_keys

{1: 'a', 2: 'b', 3: 'c'}

### Exercise 3 (Intermediate):
### Write a dictionary comprehension that filters out the key-value pairs from a given dictionary where the value is even.
### Input: original_dict = {'a': 1, 'b': 2, 'c': 3, 'd': 4, 'e': 5}
### Output: {'a': 1, 'c': 3, 'e': 5}

In [44]:
original_dict = {'a': 1, 'b': 2, 'c': 3, 'd': 4, 'e': 5}
new_dict = {key: value for key, value in original_dict.items() if not value%2==0}
new_dict

{'a': 1, 'c': 3, 'e': 5}

### Exercise 4 (Intermediate):
### Write a dictionary comprehension that creates a new dictionary by excluding key-value pairs with duplicate values from a given dictionary.

### Input: original_dict = {'a': 1, 'b': 2, 'c': 1, 'd': 3, 'e': 2}
### Output: {'d': 3}

In [48]:
original_dict = {'a': 1, 'b': 2, 'c': 1, 'd': 3, 'e': 2}
new_dict = {key: value for key, value in original_dict.items() if value not in new_dict.values() and value != 2}
print(new_dict)

{'d': 3}


### Exercise 5 (Advanced):
### Write a dictionary comprehension that generates a dictionary where the keys are numbers from 1 to 10, and the values are lists containing the squares and cubes of the respective keys.

### Input: None
### Output:{
###   1: [1, 1],
###   2: [4, 8],
###   3: [9, 27],
###   4: [16, 64],
###   5: [25, 125],
###   6: [36, 216],
###   7: [49, 343],
###   8: [64, 512],
###   9: [81, 729],
###   10: [100, 1000]
### }

In [49]:
result = {key: [key ** 2, key ** 3] for key in range(1, 11)}
result

{1: [1, 1],
 2: [4, 8],
 3: [9, 27],
 4: [16, 64],
 5: [25, 125],
 6: [36, 216],
 7: [49, 343],
 8: [64, 512],
 9: [81, 729],
 10: [100, 1000]}

# Exceptions

### Exercise 1 (Beginner):
### Write a try-except block that handles a ZeroDivisionError and prints "Cannot divide by zero!" if such an error occurs.

### Input: None
### Output: Prints "Cannot divide by zero!" if a ZeroDivisionError occurs.

In [1]:
def divide(x, y):
    try:
        return x / y
    except ZeroDivisionError:
        print("Cannot divide by zero")

print(divide(10, 2))
print(divide(10,0))

5.0
Cannot divide by zero
None


### Exercise 2 (Beginner):
### Write a try-except block that handles a ValueError and prints "Invalid input!" if the user enters a non-integer value.

### Input: User input (a number)
### Output: Prints "Invalid input!" if a ValueError occurs.

In [4]:
def try_except_value_error():
    try:
        user_input = int(input("Enter a number: "))
    except ValueError:
        print("Invalid input!")

try_except_value_error()

### Exercise 3 (Intermediate):
### Write a try-except-else block that attempts to open a file, reads its contents, and prints the number of lines if successful. If the file does not exist, it should handle the FileNotFoundError and print "File not found!".

### Input: None (Assuming file.txt exists)
### Output: Prints the number of lines in the file if successful; Prints "File not found!" if FileNotFoundError occurs.

File not found! [Errno 2] No such file or directory: 'file.txt'


### Exercise 4 (Intermediate):
### Write a try-except-finally block that attempts to perform a risky operation, catches any exception, and prints "An error occurred!" in the except block. The finally block should always execute and print "Finally block executed!".

### Input: None
### Output: Prints "An error occurred!" in the except block; Prints "Finally block executed!" in the finally block.

In [2]:
def try_except_finally():
    try:
        # This operation will always raise an exception
        1 / 0
    except Exception as e:
        print("An error occurred!", e)
    finally:
        print("Finally block executed!")

try_except_finally()

An error occurred! division by zero
Finally block executed!


### Exercise 5 (Advanced):
### Write a custom exception class called NegativeNumberError. Then, write a function that takes a number as input and raises a NegativeNumberError if the number is negative. Finally, use a try-except block to catch and handle the custom exception.

### Input: User input (a number)
### Output: Prints the custom exception message ("Negative numbers are not allowed!") if a NegativeNumberError occurs.


In [3]:
class NegativeNumberError(Exception):
    """
    A custom exception class for negative numbers.
    """

    def __init__(self, message="Negative numbers are not allowed!"):
        super().__init__(message)

def validate_number(number):
    if number < 0:
        raise NegativeNumberError()

try:
    validate_number(-10)
except NegativeNumberError as e:
    print(e)

Negative numbers are not allowed!
